### PR TITLE
Feat/US34242: Mint new NFT 

### PR DESCRIPTION
**Pull request by anshulkhare-lab49**
**Description**:
Implemented flow to mint NFT tokens
- Created Modal to mint NFT tokens and Integrated it in NFT DAO Dashboard
- Created Hook to Mint new Tokens for any NFT

Screenshot 2023-07-13 at 9 58 42 AM

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: src/dex-ui/pages/dao/MintNFTModal.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                fieldPlaceholder="Enter URL"
+                fieldLabel=""
+                fieldButtonText={fieldButtonText}
+                append={append}
+                remove={remove}
+                register={register}
+              />
+            </SimpleGrid>
+            <Button key="open-mint-nft" variant="primary" onClick={handleMintNFTClick}>
+              <Text textStyle="p small semibold">Mint NFT Tokens</Text>
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/MintNFTModal.tsx:92]: Please add `color={Color.White_02}` to `Text` to match the UI style

File: src/dex-ui/pages/dao/DashboardHeader.tsx [hunk 2]
```diff
@@ -20,11 +22,22 @@ interface DashboardHeaderProps {
...
   const navigate = useNavigate();
+  const { data: token } = useToken(govTokenId ?? "");
+  let showMintNFTButton = false;
+  if (isNFT && token) {
+    const maxSupply = token.data.max_supply;
+    const totalSupply = token.data.total_supply;
+    if (maxSupply && totalSupply) {
+      showMintNFTButton = Number(token.data.max_supply) > Number(token.data.total_supply);
+    }
+  }
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/DashboardHeader.tsx:41]: we can simplify the above condition:

` const { daoAccountId, name, type, logoUrl, govTokenId = "", safeId, isMember, isAdmin, isNFT, handleMintNFT } = props;
  const navigate = useNavigate();
  const { data: token } = useToken(govTokenId);
  const showMintNFTButton = type === DAOType.NFT && Number(token?.data.max_supply) > Number(token?.data.total_supply);`

File: src/dex-ui/pages/dao/DashboardHeader.tsx [hunk 3]
```diff
@@ -74,6 +87,9 @@ export function DashboardHeader(props: DashboardHeaderProps) {
           <Flex height="40px" alignItems="center">
             <Breadcrumb to="/daos" label="Back to DAOs" />
           </Flex>
+          <Flex height="40px" alignItems="center">
+            {showMintNFTButton && token && <MintNFTModal token={token} handleMintNFT={handleMintNFT} />}
+          </Flex>
           <Flex direction="column" gap="0.6rem">
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/DashboardHeader.tsx:93]: if `showMintNFTButton` is `false` then we are adding extra `Flex` which will be extra gap between elements in UI, 
we could do this:

`          {
            showMintNFTButton ? (
              <Flex height="40px" alignItems="center">
                {showMintNFTButton && token && <MintNFTModal token={token} handleMintNFT={handleMintNFT} />}
              </Flex>
            ) : undefined
          }
`

File: src/dex-ui/pages/dao/MintNFTModal.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  token: MirrorNodeTokenById;
+  handleMintNFT?: (tokenLinks: string[]) => void;
+}
+export const MintNFTModal = (props: MintNFTModalProps) => {
+  const { token, handleMintNFT } = props;
+  const [dialogsOpenState, setDialogsOpenState] = useState(false);
+  const defaultFieldValue = { value: "" };
+  const maxSupply = Number(token.data.max_supply);
+  const totalSupply = Number(token.data.total_supply);
+  let fieldButtonText = "+ Add Link";
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/MintNFTModal.tsx:18]: Just for Information `total_supply` is = minted + burned `tokens`

File: src/dex-ui/pages/dao/MintNFTModal.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    },
+  });
+
+  const { control, register, getValues } = mintNFTTokenForm;
+
+  const { fields, append, remove } = useFieldArray({
+    control,
+    name: "tokenLinks",
+    rules: { maxLength: 4 },
+  });
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/MintNFTModal.tsx:32]: do we need `maxLength` rule here.?

File: src/dex-ui/pages/dao/MintNFTModal.tsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+  if (totalSupply + getValues().tokenLinks.length >= maxSupply) {
+    fieldButtonText = "";
+  }
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/MintNFTModal.tsx:36]: we can simplify this `  const fieldButtonText = totalSupply + getValues().tokenLinks.length >= maxSupply ? "" : "+ Add Link";`

File: src/dex-ui/pages/dao/MintNFTModal.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            </Text>
+          </Button>
+        }
+        openDialogButtonText="Mint NFT"
+        body={
+          <Flex flexDirection="column" height="fit-content" gap="1rem">
+            <Flex direction="row" alignItems="start" justifyContent="start" flex="2" gap="1">
+              <Text textStyle="h4">Token:</Text>
+              <Text textStyle="b3">{token.data.symbol}</Text>
+            </Flex>
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/pages/dao/MintNFTModal.tsx:69]: we can use `gap` in `rem` values here.

File: src/dex-ui/services/contractsUAT.json [hunk 8]
```diff
@@ -215,11 +215,11 @@
...
-    "transparentProxyAddress": "0x0000000000000000000000000000000000e53565",
-    "transparentProxyId": "0.0.15021413"
+    "id": "0.0.15051802",
+    "address": "0x0000000000000000000000000000000000e5ac1a",
+    "timestamp": "2023-07-05T12:01:38.207Z",
+    "hash": "dfe74c0b15aa61042d372b8819cb121d",
+    "transparentProxyAddress": "0x0000000000000000000000000000000000e5ac1c",
+    "transparentProxyId": "0.0.15051804"
   }
 ]
```
Comment (roshan-bisht-lab49) [inline src/dex-ui/services/contractsUAT.json:225]: In `FormInputList` file we are rendering the button when text is empty for `fieldButtonText `  we can optimise it to 

`      {
        fieldButtonText.length > 0 ? (
          <Button paddingLeft="0.5rem" variant="link" onClick={handleAddInputClicked}>
            {fieldButtonText}
          </Button>
        ) : undefined
      }
`

**Review by roshan-bisht-lab49 — CHANGES_REQUESTED** [review]
Minor UI suggestions

File: src/dex-ui/pages/dao/DashboardHeader.tsx (hunk 2 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/DashboardHeader.tsx:41]: You are right. made the suggested changes.

File: src/dex-ui/pages/dao/DashboardHeader.tsx (hunk 3 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/DashboardHeader.tsx:93]: You are right. made the suggested changes.

File: src/dex-ui/pages/dao/MintNFTModal.tsx (hunk 1 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/MintNFTModal.tsx:92]: You are right. made the suggested changes.

File: src/dex-ui/pages/dao/MintNFTModal.tsx (hunk 5 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/MintNFTModal.tsx:32]: No we do not. removed it.

File: src/dex-ui/pages/dao/MintNFTModal.tsx (hunk 6 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/MintNFTModal.tsx:36]: You are right. made the suggested changes.

File: src/dex-ui/pages/dao/MintNFTModal.tsx (hunk 7 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/MintNFTModal.tsx:69]: When we give numbers to chakra UI compoenent. it converts it to rems behind the scene. 1unit = 0.25rem
https://chakra-ui.com/docs/styled-system/theme#spacing

File: src/dex-ui/services/contractsUAT.json (hunk 8 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/services/contractsUAT.json:225]: You are right. made the suggested changes.

File: src/dex-ui/pages/dao/MintNFTModal.tsx (hunk 4 above)
Comment (anshulkhare-lab49) [reply src/dex-ui/pages/dao/MintNFTModal.tsx:18]: I checked the flow for burned tokens. total_supply = minted - burned tokens. so if we have max limit of 10. minted 5 tokens and then burned 3, then we can mint 8 more tokens. ( 10 - (5-3)).

**Review by roshan-bisht-lab49 — APPROVED** [review]

**Pull request merged**